src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -631,7 +635,85 @@ public void AddInput(string command)
                 FinishCurrentSubmissionInput();
             }
 
-            private void AppendUncommittedInput(string text)
+            /// <summary>Implements <see cref="IInteractiveWindow2.AddToHistory"/>.</summary>
+            public void AddToHistory(string command) 
+            {
+                if (string.IsNullOrEmpty(command)) 
+                {
+                    return;
+                }
+
+                _buffer.Flush();
+
+                var sourceSpans = _projectionBuffer.CurrentSnapshot.GetSourceSpans();
+                int insertionIndex;
+                switch (State) 
+                {
+                    case State.Starting:
+                    case State.Initializing:
+                        throw new InvalidOperationException(InteractiveWindowResources.NotInitialized);
+                    case State.Resetting:
+                    case State.ResettingAndReadingStandardInput:
+                        throw new InvalidOperationException(InteractiveWindowResources.IsResettings);
+                    case State.ExecutingInput:
+                        insertionIndex = sourceSpans.Count;
+                        break;
+                    case State.ExecutingInputAndReadingStandardInput:
+                    case State.WaitingForInput:
+                    case State.WaitingForInputAndReadingStandardInput:
+                        insertionIndex = IndexOfLastSpan(sourceSpans, ReplSpanKind.Prompt);
+                        if (insertionIndex == -1) 
+                        {
+                            insertionIndex = 0;
+                        }
+                
+                        break;
+                    default:
+                        throw new InvalidOperationException();
```

Comment:
`throw ExceptionUtilities.UnexpectedValue(State);` since this is supposedly to be exhaustive.